packages/react-reconciler/src/ReactFiberHotReloading.js

```diff
@@ -0,0 +1,312 @@
+      ) {
+        // ForwardRef is special because its resolved .type is an object,
+        // but it's possible that we only have its inner render function in the map.
+        // If that inner render function is different, we'll build a new forwardRef type.
+        const currentRender = resolveFunctionForHotReloading(type.render);
+        if (type.render !== currentRender) {
+          return {...type, render: currentRender};
+        }
+      }
+      return type;
+    }
+    // Use the latest known implementation.
+    return family.currentType;
+  } else {
+    return type;
+  }
+}
+
+export function isCompatibleFamilyForHotReloading(
+  fiber: Fiber,
+  element: ReactElement,
+): boolean {
+  if (__DEV__) {
+    if (familiesByType === null) {
+      // Hot reloading is disabled.
+      return false;
+    }
+
+    const prevType = fiber.elementType;
+    const nextType = element.type;
+    // If we got here, we know types aren't === equal.
+    let needsCompareFamilies = false;
+    switch (fiber.tag) {
+      case FunctionComponent: {
+        if (typeof nextType === 'function') {
+          needsCompareFamilies = true;
+        }
+        break;
+      }
+      case ForwardRef: {
+        if (
+          typeof nextType === 'object' &&
+          nextType !== null &&
+          nextType.$$typeof === REACT_FORWARD_REF_TYPE
+        ) {
+          needsCompareFamilies = true;
+        }
+        break;
+      }
+      case MemoComponent:
+      case SimpleMemoComponent: {
+        if (
+          typeof nextType === 'object' &&
+          nextType !== null &&
+          nextType.$$typeof === REACT_MEMO_TYPE
+        ) {
+          // TODO: if it was but can no longer be simple,
+          // we shouldn't set this.
+          needsCompareFamilies = true;
+        }
+        break;
+      }
+      // TODO: maybe support lazy?
+      default:
+        return false;
+    }
+
+    // Check if both types have a family and it's the same one.
+    if (needsCompareFamilies) {
+      // Note: memo() and forwardRef() we'll compare outer rather than inner type.
+      // This means both of them need to be registered to preserve state.
+      // If we unwrapped and compared the inner types for wrappers instead,
+      // then we would risk falsely saying two separate memo(Foo)
+      // calls are equivalent because they wrap the same Foo function.
+      const prevFamily = familiesByType.get(prevType);
+      if (
+        prevFamily !== undefined &&
+        prevFamily === familiesByType.get(nextType)
+      ) {
+        return true;
+      }
+    }
+    return false;
+  } else {
+    return false;
+  }
+}
+
+export function shouldForceRenderForHotReloading(fiber: Fiber | null): boolean {
+  if (__DEV__) {
+    if (fiber === null) {
+      return false;
+    }
+    if (fibersWithForcedRender === null) {
+      // Not hot reloading now.
+      return false;
+    }
+    return fibersWithForcedRender.has(fiber);
```

Comment:
I don't think we should model these as synchronous. This should work just as well as an async operation. This is trying to cheat the async system by being clever and now that is giving me a lot of homework to figure out if that is actually safe and when it might not be.

Comparing `current.type !== workInProgress.type` solves one of the cases and works in async. Let's see if we can also solve the other one.

(Maybe we still do them sync for purposes of avoiding yielding bugs but that could trigger warnings e.g. against suspending in sync mode.)

Reply:
Actually, isn't assuming synchronous rendering already broken for deprioritized subtrees like hidden and dehydrated trees? flushSync doesn't force those into Sync priority.